server/src/main/java/org/opensearch/index/IndexService.java

```diff
@@ -454,7 +458,8 @@ public IndexService(
             s -> {},
             (shardId) -> ReplicationStats.empty(),
             clusterDefaultMaxMergeAtOnce,
-            clusterMergeSchedulerConfig
+            clusterMergeSchedulerConfig,
+            null
```

Comment:
I haven't looked at this super deeply, but a registry seems like something that would be pretty easy to implement where a default empty instance could be created so it would be guaranteed non-null and eliminate a class of possible bugs and the need for any null checking on the field.

e.g. `DataSourceRegistry.EMPTY` or something similar